### PR TITLE
[WIP] Add distributed inference for llama3.2 vision

### PR DESCRIPTION
**Pull request by acisseJZhong**
#### Context
What is the purpose of this PR? Is it to
- [ ] add a new feature
- [ ] fix a bug
- [ ] update tests and/or documentation
- [ ] other (please add here)

Please link to any issues this PR addresses.

#### Changelog
What are the changes made in this PR?
*

#### Test plan
Please make sure to do each of the following if applicable to your PR. If you're unsure about any one of these just ask and we will happily help. We also have a [contributing page](https://github.com/pytorch/torchtune/blob/main/CONTRIBUTING.md) for some guidance on contributing.

- [ ] run pre-commit hooks and linters (make sure you've first installed via `pre-commit install`)
- [ ] add [unit tests](https://github.com/pytorch/torchtune/tree/main/tests/torchtune) for any new functionality
- [ ] update [docstrings](https://github.com/pytorch/torchtune/tree/main/docs/source) for any new or updated methods or classes
- [ ] run unit tests via `pytest tests`
- [ ] run recipe tests via `pytest tests -m integration_test`
- [ ] manually run any new or modified recipes with sufficient proof of correctness
- [ ] include relevant commands and any other artifacts in this summary (pastes of loss curves, eval results, etc.)

#### UX
If your function changed a public API, please add a dummy example of what the user experience will look like when calling it.
Here is a [docstring example](https://github.com/pytorch/torchtune/blob/6a7951f1cdd0b56a9746ef5935106989415f50e3/torchtune/modules/vision_transformer.py#L285)
and a [tutorial example](https://pytorch.org/torchtune/main/tutorials/qat_finetune.html#applying-qat-to-llama3-models)

- [ ] I did not change any public API
- [ ] I have added an example to docs or docstrings

### CODE REVIEW

File: torchtune/models/llama3_2_vision/_parallelism.py [hunk 1]
```diff
@@ -0,0 +1,70 @@
...
+    "decoder.tok_embeddings.embedding": RowwiseParallel(
+        input_layouts=Replicate()
+    ),
+    "decoder.tok_embeddings.fusion_embedding": RowwiseParallel(
+        input_layouts=Replicate()
+    ),
+    "decoder.output": ColwiseParallel(
+        output_layouts=Replicate()
+    ),
+    "decoder.layers.*.attn.q_proj": ColwiseParallel(input_layouts=Replicate()),
```
Comment (SalmanMohammadi) [inline torchtune/models/llama3_2_vision/_parallelism.py:39]: very noob q: should we be sharding the kv-cache buffers too? are they automatically replicated across ranks?

Comment (acisseJZhong) [reply torchtune/models/llama3_2_vision/_parallelism.py:39]: great question! I believe the primary reason is that parallelization is typically applied during training rather than inference. Therefore most parallelism plan I saw (from Torchtitan and Torchtune) does not include the KV cache in their parallelism plans.

File: recipes/dev/generate_v2_distributed.py [hunk 2]
```diff
@@ -15,8 +15,9 @@
 
 from torchtune import config, training, utils
 from torchtune.data import load_image, Message, padded_collate_tiled_images_and_mask
-
 from torchtune.generation import sample
+from torchtune.models.llama3_2_vision import parallelize_llama3_2_vision
```
Comment (pbontrager) [inline recipes/dev/generate_v2_distributed.py:19]: A recipe shouldn't have model specific imports

File: recipes/dev/generate_v2_distributed.py [hunk 3]
```diff
@@ -105,12 +106,18 @@ def setup(self, cfg: DictConfig) -> None:
...
-        parallelize_module(
-            model,
-            tp_device_mesh,
-            parallelize_plan=config.instantiate(cfg.parallelize_plan),
-        )
+        model = training.prepare_mha_for_tp(model, tp_device_mesh)
+
+        # Special handling for 3.2 vision model, we need to parallelize the model differently as wildcard is not
+        # working for now. Need to change back to wildcard once the issue is fixed.
+        if isinstance(model, DeepFusionModel):
```
Comment (pbontrager) [inline recipes/dev/generate_v2_distributed.py:113]: This logic should be pushed into the "parallelize_module" utility and not exposed in the recipe.

File: torchtune/models/llama3_2_vision/_parallelism.py [hunk 4]
```diff
@@ -0,0 +1,141 @@
...
+    RowwiseParallel,
+)
+from torch.distributed.tensor.parallel.style import ParallelStyle
+from torchtune.modules import TransformerSelfAttentionLayer
+from torchtune.modules.model_fusion import FusionLayer
+
+
+# Define the Tensor Parallel plan for Llama3.2 vision model
+# TODO: wildcard is not working for the 3.2 deep fusion model. Use `parallelize_llama3_2_vision` for now.
+LLAMA3_2_VISION_TP_PLAN = {
```
Comment (pbontrager) [inline torchtune/models/llama3_2_vision/_parallelism.py:25]: The encoder doesn't get parallelized?

File: torchtune/models/llama3_2_vision/_parallelism.py [hunk 5]
```diff
@@ -0,0 +1,141 @@
+# Copyright (c) Meta Platforms, Inc. and affiliates.
```
Comment (pbontrager) [inline torchtune/models/llama3_2_vision/_parallelism.py:1]: Since this is a patch to make deepfusion models work until the wildcard issue is fixed, I'd be inclined to move this out of llama3.2 and directly into the distributed file to get called when given a DeepFusion model within parallelize_model. It could be kept at the bottom of the file after some "###### TITLE #####" style separator

File: torchtune/training/_distributed.py [hunk 6]
```diff
@@ -299,16 +298,15 @@ def load_from_full_model_state_dict(
                     ),
                     requires_grad=sharded_meta_param.requires_grad,
                 )
-
-            elif not hasattr(sharded_meta_param, "device_mesh"):
-                # In cases where parts of the model aren't sharded, some parameters will be plain tensors
-                sharded_tensor = full_tensor
-            else:
+            elif isinstance(sharded_meta_param, DTensor):
```
Comment (pbontrager) [inline torchtune/training/_distributed.py:301]: I don't remember why we were using device_mesh to infer DTensor's but this code cleanup does make me nervous that there was an edge case where checking for DTensor correctly didn't work. Are you sure there's no such case?

File: torchtune/training/_distributed.py [hunk 7]
```diff
@@ -624,4 +622,6 @@ def prepare_mha_for_tp(
             m.num_heads = m.num_heads // tp_size
             m.num_kv_heads = m.num_kv_heads // tp_size
             m.embed_dim = m.embed_dim // tp_size
+    if is_fusion_model:
```
Comment (pbontrager) [inline torchtune/training/_distributed.py:625]: So before we were only returning the decoder and now we return the full model? Was this a bug?

**Review by pbontrager — COMMENTED** [review]
Thanks for enabling this. I left a few comments around where we want to put and expose these changes.

File: torchtune/models/llama3_2_vision/_parallelism.py (hunk 4 above)
Comment (acisseJZhong) [reply torchtune/models/llama3_2_vision/_parallelism.py:25]: yes, I think encoder doesn't needs to be parallelized as it's not that big compared to decoder layers

File: torchtune/training/_distributed.py (hunk 6 above)
Comment (acisseJZhong) [reply torchtune/training/_distributed.py:301]: it actually took a while for me to debug this; https://github.com/pytorch/pytorch/blob/main/torch/distributed/tensor/_api.py#L239 DTensor doesn't seems to have this device_mesh field, so this elif block was never activated. 

Curious why we can't just do isinstance DTensor?

File: torchtune/training/_distributed.py (hunk 7 above)
Comment (acisseJZhong) [reply torchtune/training/_distributed.py:625]: this is already fixed on the public repo already i believe. I need to rebase.

File: torchtune/models/llama3_2_vision/_parallelism.py (hunk 5 above)
Comment (acisseJZhong) [reply torchtune/models/llama3_2_vision/_parallelism.py:1]: are you suggesting moving this hack logic into _distributed.py? I prefer keeping it here as it's 3.2 model specific?

File: recipes/dev/generate_v2_distributed.py (hunk 3 above)
Comment (acisseJZhong) [reply recipes/dev/generate_v2_distributed.py:113]: yeah I can have another `parallelize_module` function in `_distributed.py`. But that wouldn't solve the problem(having this if else check)? it just makes the recipe looks cleaner.

File: torchtune/_recipe_registry.py [hunk 8]
```diff
@@ -449,6 +449,10 @@ class Recipe:
                 name="llama3_1/70B_generation_distributed",
                 file_path="llama3_1/70B_generation_distributed.yaml",
             ),
+            Config(
+                name="llama3_2/90B_generation_distributed",
+                file_path="llama3_2/90B_generation_distributed.yaml",
```
Comment (ebsmothers) [inline torchtune/_recipe_registry.py:454]:
```suggestion
                name="llama3_2_vision/90B_generation_distributed",
                file_path="llama3_2_vision/90B_generation_distributed.yaml",
```

File: recipes/dev/generate_v2_distributed.py (hunk 2 above)
Comment (ebsmothers) [reply recipes/dev/generate_v2_distributed.py:19]: Bumping this comment, can we take this out?

File: torchtune/models/llama3_2_vision/_parallelism.py (hunk 5 above)
Comment (ebsmothers) [reply torchtune/models/llama3_2_vision/_parallelism.py:1]: To cast a tiebreaking vote, I also prefer to keep it here. (I don't wanna add even more code to `_distributed.py` than we have already)

File: recipes/dev/generate_v2_distributed.py (hunk 3 above)
Comment (ebsmothers) [reply recipes/dev/generate_v2_distributed.py:113]: Yeah the if/else is not going to go away. But we shouldn't put it in the recipe (more boilerplate that the user shouldn't have to care about)

File: torchtune/training/_distributed.py (hunk 6 above)
Comment (ebsmothers) [reply torchtune/training/_distributed.py:301]: Sorry I'm also a bit confused by this.. what is the actual change that's happening here? I think DTensor does have this field, see [here](https://github.com/pytorch/pytorch/blob/main/torch/distributed/tensor/_api.py#L580). Is there some case where it's null?

Comment (acisseJZhong) [reply torchtune/training/_distributed.py:301]: Yeah we when using the `not hasattr(sharded_meta_param, "device_mesh")`, its not entering this block. I changed the logic to using `isinstance(DTesnor)`

File: recipes/configs/llama3_2_vision/90B_generation_distributed.yaml [hunk 9]
```diff
@@ -0,0 +1,51 @@
...
+#    tune run --nproc_per_node 8 dev/generate_v2_distributed --config llama3_2_vision/90B_generation_distributed
+
+output_dir: ./
+
+# Model arguments
+model:
+  _component_: torchtune.models.llama3_2_vision.llama3_2_vision_90b
+
+parallelize_plan:
+  _component_: torchtune.models.llama3_2_vision.llama3_2_vision_tp_plan
```
Comment (ebsmothers) [inline recipes/configs/llama3_2_vision/90B_generation_distributed.yaml:17]: I know it's due to the wildcard issue but it's confusing to expose this in the config and then use `parallelize_llama3_2_vision` in the recipe. We should simplify this (personally I would just rip out the wildcards until they're actually supported, but up to you)

**Review by ebsmothers — COMMENTED** [review]
A few more comments, the main thing is to understand why the fix in `load_from_full_model_state_dict` is the correct one. Two other general things: 

(1) Can you share tok/s and peak memory numbers in the PR summary? 
(2) Does the job actually shutdown after generation completes? At least for me it does the generation and then just hangs until NCCL timeout. (Maybe it's a problem unique to my machine, but just want to confirm.)
